### PR TITLE
Support partial test runs on local single-backend environments

### LINKED ISSUE #366: Allow running local tests with a subset of ML back-ends
Currently, tests fail altogether unless all back-ends are installed. That's good for the CI, but people contributing a small bugfix probably won't use the all-back-ends docker, so they can't run the tests locally before they make a PR. We could add a check in the test logic that tests which back-ends are available, which can fail with a message like 

> Not all back-ends are available, meaning tests will fail. To skip tests that require unavailable back-ends, use `pytest --skip-unavailable-backends`.

Then simply skip any test that imports functions from a specific back-end. That way the CI pipeline will still fail if we mess something up in a docker update, but contributors can conveniently run tests locally.

### LINKED ISSUE #369: Make tests pass with non-tensorflow back-ends
A number of regression tests now rely on hardcoded return values from a consistent seed, but these seeds are not actually harmonized across back-ends. For example, [test_covariance_sampling_lines()](https://github.com/tue-bmd/zea/blob/4cbd6a3117ecbdbf889c86c103baffd7fdb2e423/tests/test_agent.py#L162) fails with jax because of this.

### LINKED ISSUE #375: Device tests do not pass local pytest run
Image

Tests are skipped for the runners, but some actually fail.

This could be the openh-rf branch #358 , main branch, or both, depending on when you look at this issue.

### LINKED ISSUE #401: Tests fail on jax due to stateful random seeds
The following tests currently fail on jax, all due to random seed-related issues:

```
tests/test_augmentations.py
tests/test_flow_matching.py
tests/test_generative.py
```

This was originally fixed in #373, but that made the PR too big to review, so I'm splitting off those changes into a separate PR.

### PR DESCRIPTION
**Pull request by Louisvh**
This adds markers for tests to exclude them on specific back-ends. When that back-end is not available, the test is skipped. It currently also skips any tests that do a multi-backend equality checks.

Tests can be either marked as entirely backend-specific:

```python
@pytest.mark.tensorflow
def test_smsle():
    """Test SMSLE loss function"""
    from zea.backend.tensorflow.losses import SMSLE
    ...
```

or only part of the test can be marked:

```python
def test_foo():
    ...
    # computation, assorted asserts
    ...
    if runs_on("tensorflow"):
        import tensorflow as tf

        expected_value = tf.image.ssim( ... )
        assert value == expected_value
```

Currently tests pass locally on jax. We'll probably want to integrate a jax-run in the CI as well, because I already ran into a jax-specific bug in the augmentation tests. 

## Summary by CodeRabbit

* **New Features**
  * Tests now dynamically detect the available ML backends, skip missing-backend cases automatically, and support new CLI flags to skip unavailable backends and override Torch usage.
  * Notebook test discovery is improved, including TensorFlow-specific notebook selection.
* **Bug Fixes**
  * Backend-conditional assertions for metrics and regressions to prevent false failures on unsupported backends.
  * Seed handling updates improve reproducibility and correctness across backends and symbolic batches.
* **Tests**
  * Added broader cross-backend coverage for tensor split seeding and TensorFlow symbolic augmentation behavior.
* **Documentation**
  * Expanded contributing guide with light/heavy/notebook test instructions and exact pytest markers.

### CODE REVIEW

**Comment by tristan-deep** [comment]
Very useful! Should make local testing easier. 

One thing I'm thinking is that for the tests you see are failing for JAX right now, we simply add the decorator there too to run on all backends (or now with your update possibly the backends that are available).

**Comment by tristan-deep** [comment]
@Louisvh you can put the PR in draft mode if it is not ready and remove the WIP from the title :)

**Comment by Louisvh** [comment]
> @Louisvh you can put the PR in draft mode if it is not ready and remove the WIP from the title :)

Yeah, I only realised I forgot to mark it as draft after I'd gotten about a hundred notifications from coderabbit 😅.

I'd started with trying to auto detect back-end specific tests, but it probably makes more sense to get rid of that and just keep the decorators. It's not _that_ many tests, and simplicity is probably more important than convenience here.

**Comment by wesselvannierop** [comment]
I realize now that you can quite easily install CPU-only versions of all backends ofcourse. The github runners do the same.

**Comment by Louisvh** [comment]
All tests now pass on jax, apart from the ones mentioned in #401. A fix for those errors is split off into #402 for easier review.

**Comment by tristan-deep** [comment]
This is really useful, thanks @Louisvh !! At first glance looks good. Will test after all the #358 is done, hopefully this week.

**Comment by Louisvh** [comment]
TODO: cross-reference these changes with the problem mentioned in #375.

**Comment by tristan-deep** [comment]
@Louisvh as promised would get back to you. Could you reserve merge conflicts? #358 is now merged so can test your PR after.

**Comment by tristan-deep** [comment]
Maybe related (as @Louisvh suggested), #458 now eradicates all loose ML imports throughout the codebase.

**Comment by tristan-deep** [comment]
Hi Louis! Great changes.

Added the following in my latest commit:
- notebook print fixed and actually counts number of notebooks that will be run for your command
- contributing.rst updated with new pytest options
- prints warning message how many tests are skipped when using `--skip-unavailable-backends`

I tested on my mac with:

```bash
uv sync --extra dev
uv add jax
source .venv/bin/activate

pytest -m 'not heavy and not notebook' --skip-unavailable-backends
```

This is an environment deliberately only installed with jax, still some tests are failing (and not skipped / fixed), this might be solved in #402 as I understand @Louisvh ?

See output log below:

<details closed>
<summary>Pytest log</summary>
```bash
tests/data/test_app.py ...........................                                                                                                [  1%]
tests/data/test_conversion_scripts.py ......F........................................                                                             [  5%]
tests/data/test_data_format.py ..................................                                                                                 [  7%]
tests/data/test_dataloader.py ..............................F.......................                                                              [ 11%]
tests/data/test_dataset.py ...........................                                                                                            [ 13%]
tests/data/test_file.py ......................................................................................................................... [ 22%]
                                                                                                                                                  [ 22%]
tests/data/test_file_operations.py .....................                                                                                          [ 24%]
tests/data/test_process.py .............                                                                                                          [ 25%]
tests/data/test_spec.py ......................................................................................................................... [ 34%]
..........................                                                                                                                        [ 36%]
tests/test_agent.py ...................                                                                                                           [ 37%]
tests/test_augmentations.py FFFFFFFFFF.Fss                                                                                                        [ 38%]
tests/test_autograd.py ss...                                                                                                                      [ 39%]
tests/test_backend.py s.Fs.ssss.F.                                                                                                                [ 39%]
tests/test_beamformer.py sssssssssssssssssssss                                                                                                    [ 41%]
tests/test_caching.py .........                                                                                                                   [ 42%]
tests/test_config_validation.py .........s.............                                                                                           [ 43%]
tests/test_configs.py ...................................................                                                                         [ 47%]
tests/test_custom_ops.py ...............                                                                                                          [ 48%]
tests/test_device.py .............ssssssssssssssssssssssssssssssssssssss....ssss..s....sss                                                        [ 53%]
tests/test_display.py sssss.ssssssssss...............                                                                                             [ 56%]
tests/test_flow_matching.py ..........FFF.s.F....F..                                                                                              [ 57%]
tests/test_fnumber_mask.py ...                                                                                                                    [ 58%]
tests/test_generate_keras_ops.py ............                                                                                                     [ 58%]
tests/test_generative.py ...ss.FFF....                                                                                                            [ 59%]
tests/test_gumbel.py ....                                                                                                                         [ 60%]
tests/test_helpers.py ss.                                                                                                                         [ 60%]
tests/test_imports.py F.ss...                                                                                                                     [ 60%]
tests/test_io_lib.py ........                                                                                                                     [ 61%]
tests/test_keras_ops.py ..                                                                                                                        [ 61%]
tests/test_main.py ..........                                                                                                                     [ 62%]
tests/test_metrics.py ssssssssssssssss..                                                                                                          [ 63%]
tests/test_model_loading.py ....                                                                                                                  [ 64%]
tests/test_models.py ..EE.E                                                                                                                       [ 64%]
tests/test_object.py ..                                                                                                                           [ 64%]
tests/test_operations.py sssssssssss.........sssssssss.sssssssssss.ssss.ssssss.ssssssss                                                           [ 69%]
tests/test_ops_infra.py ................................................................................                                          [ 75%]
tests/test_parameters.py .........................                                                                                                [ 76%]
tests/test_pfield.py ..                                                                                                                           [ 77%]
tests/test_pixelgrid.py .........................                                                                                                 [ 78%]
tests/test_probes.py ...........                                                                                                                  [ 79%]
tests/test_refocus.py ...............ss                                                                                                           [ 80%]
tests/test_registration.py .....                                                                                                                  [ 81%]
tests/test_scan.py ..........................................                                                                                      [ 84%]
tests/test_selection_tool.py .....                                                                                                                 [ 84%]
tests/test_tensor_ops.py ssssssss..ssssssssssssssssssssssssss.ssssssssssssssssssssssss.ss..............ssssssssssssssssssss..s..                   [ 92%]
tests/test_tracking.py ........................                                                                                                    [ 94%]
tests/test_transmit_schemes.py .                                                                                                                   [ 94%]
tests/test_unet.py ..........                                                                                                                      [ 94%]
tests/test_user_settings.py .....                                                                                                                  [ 95%]
tests/test_utils.py ...................                                                                                                            [ 96%]
tests/test_visualize.py ..................................                                                                                         [ 99%]
tests/tools/test_hf.py ...F......                                                                                                                  [100%]
=================================== 25 failed, 1080 passed, 248 skipped, 44 deselected, 3 errors in 173.68s (0:02:53) ====================================
```
</details>

EDIT: now testing with tensorflow installed as well...

**Comment by Louisvh** [comment]
I think a few of the failures might actually be from only jax being installed without CUDA; I don't think I tested that one, so something might still be mishandled there (and also not yet fixed in the follow-up pr).

**Comment by tristan-deep** [comment]
@coderabbitai full review

**Comment by tristan-deep** [comment]
Think I fixed all remaining issues! Tested with both only "tensorflow" and only "jax" installations with the `--skip-unavailable-backends` flag (with CPU only on mac) and passes all the tests! (or skips them properly now, with a nice warning).

File: zea/func/__init__.py [hunk 1]
```diff
@@ -24,10 +24,11 @@
     gaussian_filter1d,
     images_to_patches,
     interpolate_data,
-    is_jax_prng_key,
+    is_jax_key,
     is_monotonic,
     linear_sum_assignment,
     map_indices_for_interpolation,
+    materialize_seed,
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply zea/func/__init__.py:31]: don't need backwards compat for this internal func.

File: tests/test_tensor_ops.py [hunk 2]
```diff
@@ -945,3 +951,125 @@ def test_nonzero(array, size, fill_value):
...
+    with pytest.raises(TypeError):
+        split_seed(arr, 2)
+
+
+def test_split_seed_rejects_unknown_seed_type():
+    """An unsupported seed type is rejected on any backend."""
+    from zea.func.tensor import split_seed
+
+    with pytest.raises((TypeError, AssertionError)):
+        split_seed("not-a-seed", 2)
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply tests/test_tensor_ops.py]: done

File: tests/test_augmentations.py [hunk 3]
```diff
@@ -268,3 +269,44 @@ def test_random_circle_inclusion_with_height_width_ranges():
...
+    """x_is_symbolic_tensor=True + randomize_location_across_batch=True uses ops.map."""
+    import tensorflow as tf
+
+    images = np.zeros((4, 28, 28), dtype=np.float32)
+    layer = RandomCircleInclusion(radius=5, fill_value=1.0, circle_axes=(1, 2), with_batch_dim=True)
+
+    # input_signature forces a dynamic (None) batch dimension to trigger the symbolic-tensor branch.
+    @tf.function(input_signature=[tf.TensorSpec(shape=(None, 28, 28), dtype=tf.float32)])
+    def run(x):
+        return layer(x)
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply tests/test_augmentations.py]: done

File: tests/backend_utils.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    """Return whether the current test backend is not one of the excluded backends.
+    Misses are counted so pytest can report how many backend-guarded blocks were skipped.
+    """
+    import keras
+
+    active_backend = keras.backend.backend()
+    if active_backend not in backends:
+        return True
+    _record_backend_guard_skip(active_backend, backends)
+    return False
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply tests/backend_utils.py]: done

**Review by tristan-deep — APPROVED** [review]

**Comment by tristan-deep** [comment]
Added #466 to indicate that pytest with torch backend only is not supported currently.
